pinecone/index_connection.go

```diff
@@ -0,0 +1,358 @@
+	config := &tls.Config{}
+	target := fmt.Sprintf("%s:443", host)
+	conn, err := grpc.Dial(
+		target,
+		grpc.WithTransportCredentials(credentials.NewTLS(config)),
+		grpc.WithAuthority(target),
+		grpc.WithBlock(),
+	)
+
+	if err != nil {
+		log.Fatalf("fail to dial: %v", err)
+		return nil, err
+	}
+
+	dataClient := data.NewVectorServiceClient(conn)
+
+	idx := IndexConnection{apiKey: apiKey, dataClient: &dataClient, grpcConn: conn}
+	return &idx, nil
+}
+
+func (idx *IndexConnection) Close() error {
+	err := idx.grpcConn.Close()
+	return err
+}
+
+type UpsertVectorsRequest struct {
+	Vectors   []*Vector
+	Namespace string
+}
+
+func (idx *IndexConnection) UpsertVectors(ctx *context.Context, in *UpsertVectorsRequest) (uint32, error) {
+	vectors := make([]*data.Vector, len(in.Vectors))
+	for i, v := range in.Vectors {
+		vectors[i] = vecToGrpc(v)
+	}
+
+	req := &data.UpsertRequest{
+		Vectors:   vectors,
+		Namespace: in.Namespace,
+	}
+
+	res, err := (*idx.dataClient).Upsert(idx.akCtx(*ctx), req)
+	if err != nil {
+		return 0, err
+	}
+	return res.UpsertedCount, nil
+}
+
+type FetchVectorsRequest struct {
+	Ids       []string
+	Namespace string
+}
+
+type FetchVectorsResponse struct {
+	Vectors   map[string]*Vector
+	Namespace string
+	Usage     *Usage
+}
+
+func (idx *IndexConnection) FetchVectors(ctx *context.Context, in *FetchVectorsRequest) (*FetchVectorsResponse, error) {
+	req := &data.FetchRequest{
+		Ids:       in.Ids,
+		Namespace: in.Namespace,
+	}
+
+	res, err := (*idx.dataClient).Fetch(idx.akCtx(*ctx), req)
+	if err != nil {
+		return nil, err
+	}
+
+	vectors := make(map[string]*Vector, len(res.Vectors))
+	for id, vector := range res.Vectors {
+		vectors[id] = toVector(vector)
+	}
+
+	return &FetchVectorsResponse{
+		Vectors:   vectors,
+		Namespace: res.Namespace,
+		Usage:     toUsage(res.Usage),
+	}, nil
+}
+
+type ListVectorsRequest struct {
+	Prefix          *string
+	Limit           *uint32
+	PaginationToken *string
+	Namespace       string
+}
+
+type ListVectorsResponse struct {
+	VectorIds           []*string
+	Namespace           string
+	Usage               *Usage
+	NextPaginationToken *string
+}
+
+func (idx *IndexConnection) ListVectors(ctx *context.Context, in *ListVectorsRequest) (*ListVectorsResponse, error) {
+	req := &data.ListRequest{
+		Prefix:          in.Prefix,
+		Limit:           in.Limit,
+		PaginationToken: in.PaginationToken,
+		Namespace:       in.Namespace,
+	}
+	res, err := (*idx.dataClient).List(idx.akCtx(*ctx), req)
+	if err != nil {
+		return nil, err
+	}
+
+	vectorIds := make([]*string, len(res.Vectors))
+	for i := 0; i < len(res.Vectors); i++ {
+		vectorIds[i] = &res.Vectors[i].Id
+	}
+
+	return &ListVectorsResponse{
+		VectorIds:           vectorIds,
+		Namespace:           res.Namespace,
+		Usage:               &Usage{ReadUnits: res.Usage.ReadUnits},
+		NextPaginationToken: toPaginationToken(res.Pagination),
+	}, nil
+}
+
+type QueryByVectorRequest struct {
```

Comment:
To be really explicit I would call this `QueryByVectorValuesRequest` and the other `QueryByVectorIdRequest`. In both cases, the query needs a vector to compute similarity.

Reply:
f354232